### PR TITLE
Fix javascript number formatting.

### PR DESCRIPTION
**Pull request by mbrandonw**

### CODE REVIEW

**Comment by mbrandonw** [comment]
just pushed some code to validate subscription quantity, and that led me down a lil refactor path. will make some notes...

File: Makefile [hunk 1]
```diff
@@ -16,6 +16,8 @@ init-db:
 deinit-db:
 	psql template1 < database/deinit.sql
 
+reset-db: deinit-db init-db
```
Comment (mbrandonw) [inline Makefile]: this can be useful

File: Sources/PointFree/Pricing.swift [hunk 2]
```diff
@@ -523,27 +524,32 @@ private func tabStyles(
         <> selectedStyles
 }
 
-private func redirectCurrentSubscribers<A>(
```
Comment (mbrandonw) [inline Sources/PointFree/Pricing.swift]: made this reusable, new version down below...

File: Sources/PointFree/Routers/PointFreeRoutes.swift [hunk 3]
```diff
@@ -22,7 +22,7 @@ public enum Route: DerivePartialIsos {
   case launchSignup(EmailAddress)
   case login(redirect: String?)
   case logout
-  case pricing(String?, Int?)
+  case pricing(Pricing?)
```
Comment (mbrandonw) [inline Sources/PointFree/Routers/PointFreeRoutes.swift]: we can now use a first class `Pricing` value

File: Sources/PointFree/Routers/PointFreeRoutes.swift [hunk 4]
```diff
@@ -210,7 +210,10 @@ private let routers: [Router<Route>] = [
     <¢> get %> lit("logout") <% end,
 
   .pricing
-    <¢> get %> lit("pricing") %> queryParam("plan", opt(.string)) <%> queryParam("quantity", opt(.int)) <% end,
+    <¢> get %> lit("pricing")
+    %> (queryParam("plan", opt(.string)) <%> queryParam("quantity", opt(.int)))
+      .map(PartialIso.pricing >>> Optional.iso.some)
```
Comment (mbrandonw) [inline Sources/PointFree/Routers/PointFreeRoutes.swift]: here `PartialIso.pricing` is an iso `(String?, Int?) -> Pricing`, and by mapping with `PartialIso.pricing >>> Optional.iso.some` we guarantee that this succeeds in parsing even if parsing the pricing value fails.

`Optional.iso.some` is a really bad name, it should be moved to an extension on `PartialIso where B == A?`, but we can clean up later.

File: Sources/PointFree/Routers/PointFreeRoutes.swift [hunk 5]
```diff
@@ -277,3 +282,24 @@ private func verify(payload: MailgunForwardPayload) -> Bool {
...
+  fileprivate static var pricing: PartialIso {
+    return PartialIso(
+      apply: { plan, quantity in
+        let pricing: Pricing
+        if let quantity = quantity {
+          pricing = .team(quantity)
+        } else if let plan = plan, let billing = Pricing.Billing(rawValue: plan) {
+          pricing = .individual(billing)
+        } else {
+          pricing = .default
```
Comment (mbrandonw) [inline Sources/PointFree/Routers/PointFreeRoutes.swift]: I moved this logic from the site middleware to here.

now, I would have expected that this `.default` would not be necessary but it is :/ I need to dig into this more, but going to leave for now

File: Tests/PointFreeTests/SubscribeTests.swift [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import HtmlPrettyPrint
+import HttpPipeline
+@testable import PointFree
+import PointFreeTestSupport
+import Prelude
+import Optics
+import SnapshotTesting
+import XCTest
+
+final class SubscribeTests: TestCase {
```
Comment (mbrandonw) [inline Tests/PointFreeTests/SubscribeTests.swift]: I dont think we are testing the subscribe endpoint at all right now, so I added a bunch of tests

File: Sources/PointFree/Subscribe.swift [hunk 7]
```diff
@@ -11,42 +11,70 @@ public struct SubscribeData: Codable {
 }
 
 let subscribeMiddleware =
-  filterMap(require2 >>> pure, or: loginAndRedirect)
+  filterMap(
+    require1 >>> pure,
+    or: redirect(to: .pricing(nil), headersMiddleware: flash(.error, "Error creating subscription!"))
```
Comment (mbrandonw) [inline Sources/PointFree/Subscribe.swift]: this filter map makes sure there is `SubscribeData` present...

File: Sources/PointFree/Subscribe.swift [hunk 8]
```diff
@@ -11,42 +11,70 @@ public struct SubscribeData: Codable {
...
+    )
+    <<< filter(
+      get1 >>> ^\.pricing >>> validateQuantity,
+      or: redirect(to: .pricing(nil), headersMiddleware: flash(.error, "An invalid subscription quantity was used."))
```
Comment (mbrandonw) [inline Sources/PointFree/Subscribe.swift]: then this filter map makes sure the quantity is correct

File: Sources/PointFree/Subscribe.swift [hunk 9]
```diff
@@ -11,42 +11,70 @@ public struct SubscribeData: Codable {
...
+    )
+    <<< redirectActiveSubscribers(user: get2)
```
Comment (mbrandonw) [inline Sources/PointFree/Subscribe.swift]: this we redirect active subscribers away

File: Sources/PointFree/Subscribe.swift [hunk 10]
```diff
@@ -11,42 +11,70 @@ public struct SubscribeData: Codable {
...
+    <<< filterMap(require2 >>> pure, or: loginAndRedirectToPricing)
```
Comment (mbrandonw) [inline Sources/PointFree/Subscribe.swift]: and finally we require the user to be present

File: Tests/PointFreeTests/SubscribeTests.swift [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    super.tearDown()
+    AppEnvironment.pop()
+  }
+
+  func testNotLoggedIn_IndividualMonthly() {
+    let conn = connection(from: request(to: .subscribe(.some(.individualMonthly))))
+      |> siteMiddleware
+      |> Prelude.perform
+
+    #if !os(Linux)
```
Comment (mbrandonw) [inline Tests/PointFreeTests/SubscribeTests.swift]: had to disable linux tests cause snapshots of dictionaries getting reordered :/

```diff
−  pricing[individual]=monthly&token=stripe-deadbeef
+  token=stripe-deadbeef&pricing[individual]=monthly

```

File: Sources/PointFree/SiteMiddleware.swift [hunk 12]
```diff
@@ -157,17 +157,8 @@ private func render(conn: Conn<StatusLineOpen, T3<Database.Subscription?, Databa
...
-        pricing = .team(quantity)
-      } else if let plan = plan, let billing = Pricing.Billing(rawValue: plan) {
-        pricing = .individual(billing)
-      } else {
-        pricing = .default
-      }
-
-      return conn.map(const(user .*. pricing .*. route .*. unit))
+    case let .pricing(pricing):
+      return conn.map(const(user .*. (pricing ?? .default) .*. route .*. unit))
```
Comment (stephencelis) [inline Sources/PointFree/SiteMiddleware.swift]: Nice cleanup!

**Pull request merged**